### ISSUE TITLE
Option for cursor beyond EOL (non wrapping cursor)

### ISSUE THREAD

**Issue by 0x7CFE** [zed-industries/zed#14297]
### Check for existing issues

- [X] Completed

### Describe the feature

I'd like to have an editor option that allows moving the cursor caret beyond the line end in a non-wrapping manner. 

This behavior was the standard for early editors like DOS TurboC/Pascal, Delphi/CBuilder and even KDevelop/Katepart, but then KDE ditched that option for no good reason.

Being a programmer for ~30 years I am so used to this behavior that the modern wrapping cursor makes me sick. Please consider an option to bring this behavior back :pray:

**Comment by notpeter** [comment]
Are there other editors that support this and if so what is this mode/setting called?

Brainstorming:

- Is it just for navigation? If you start typing with a cursor beyond EOL does it jump back to the EOL or does it insert spaces (tabs?) between EOL and your cursor?
- Column indicator in the status bar: needs to show your current cursor position which is outside of the bounds of a 
- `ctrl-e` or `end` -- Would this jump you **back** to the end of the line if you're beyond the EOL?
- Can you click your cursor anywhere with the mouse too?
- How do selections work? What happens if you press `shift-left` while out beyond the EOL. Can you select the empty area?

**Comment by 0x7CFE** [comment]
> Are there other editors that support this and if so what is this mode/setting called?

Absolutely. This was the default for old-ish editors in DOS. For a long time it was one of the most noticeable differences between code editor and text editor. Delphi still does that to this day. It's called "wrapping cursor" or "cursor beyon EOL" (at least that how it was called in KDE).

Some relevant discussons regarding the issue:

- https://forum.sublimetext.com/t/how-cursor-beyond-end-of-line/8684
- https://stackoverflow.com/questions/17977490/how-do-i-disable-cursor-beyond-end-of-line
- Long time ago (circa KDE 3.3) it was supported in Katepart, so Kate and KDevelop used that

> * Is it just for navigation? If you start typing with a cursor beyond EOL does it jump back to the EOL or does it insert spaces (tabs?) between EOL and your cursor?

IIRC it act as if there are (virtual) spaces in between. You can navigate between actual EOL and virtual one using End key much like this happens with Home and first non-whitespace character.

> Can you click your cursor anywhere with the mouse too?

Yes, you can. Check [this video](https://youtu.be/r6bww1_nWrg?t=65) where the guy used his mouse to move cursor beyond EOL.

> How do selections work? What happens if you press shift-left while out beyond the EOL. Can you select the empty area?

I believe virtual spaces remain virtual unless used types something, so they would not be selected.

**Comment by BrutPitt** [comment]
*"Virtual Space"* (or *"caret after end line"*) **finally** is about to be implemented also in VS-Code (and only thanks to a great job and "independent" pull initiative of @x17jiri) after several years of (un-heared) requests to VSCode team:
https://github.com/microsoft/vscode/pull/228680
https://github.com/microsoft/vscode/issues/13960

**About this, there is already a full and perfectly working "unofficial" fork release: x17-Code*

Anyway it is also implemented in VisualStudio IDE editor (from ever) and in all JetBrains IDE editors (from the beginning), not (yet) in Fleet.
Below the JetBrains option setting:
![Screenshot at 2024-12-06 23-52-37](https://github.com/user-attachments/assets/de6fcf69-d27a-4875-b8af-871650a88406)

Personally I consider it an essential option.

**Comment by JosephTLyons** [comment]
Hello,

We’re closing this issue as part of a cleanup of older feature requests.

Going forward, new feature requests start as GitHub Discussions and only become Issues once the direction is agreed on. This helps ensure that Issues represent scoped, ready-to-build work and makes it easier for contributors to find work that’s ready to pick up. You can read more in our [Feature Request Guidelines](https://github.com/zed-industries/zed/discussions/47963).

As part of this transition, we’re closing feature request Issues that are older than 6 months and have fewer than 10 upvotes.

If this feature hasn’t been implemented and is still important to you, we encourage you to open a new feature request Discussion [here](https://github.com/zed-industries/zed/discussions/new?category=feature-requests). That’s where we track and evaluate new requests.

Thanks for helping shape the future of Zed.

**Comment by eugenesvk** [comment]
The last comment has 11 reactions, not sure why they didn't also upvote the original issue, but still seems to indicate the interest is higher than the threshold
